### ISSUE TITLE
[Phase 5] Spike Clustering

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/neuronal-spike-sorting#6]
## Objective

Cluster spikes into putative single units using HDBSCAN and K-means.

## Dependencies

- Phase 4: Feature Extraction (requires spike features)

## Tasks

- [ ] Implement `src/clustering/kmeans_clustering.py` for K-means
- [ ] Implement `src/clustering/hdbscan_clustering.py` for HDBSCAN
- [ ] Implement `src/clustering/cluster_utils.py` for merge/split utilities
- [ ] Compare clustering methods
- [ ] Handle outlier spikes (noise cluster)
- [ ] Save cluster assignments to `models/spike_clusters.npz`
- [ ] Write unit tests

## Files to Create

| File | Purpose |
|------|---------|
| `src/clustering/__init__.py` | Subpackage init |
| `src/clustering/kmeans_clustering.py` | K-means sorter |
| `src/clustering/hdbscan_clustering.py` | HDBSCAN sorter |
| `src/clustering/cluster_utils.py` | Merge/split tools |
| `tests/test_clustering.py` | Clustering tests |

## Starter Code

```python
# src/clustering/hdbscan_clustering.py
"""HDBSCAN-based spike clustering."""

import numpy as np
import hdbscan

class HDBSCANSorter:
    """Spike sorter using HDBSCAN clustering."""

    def __init__(
        self,
        min_cluster_size: int = 20,
        min_samples: int = 5,
        cluster_selection_epsilon: float = 0.0
    ):
        self.clusterer = hdbscan.HDBSCAN(
            min_cluster_size=min_cluster_size,
            min_samples=min_samples,
            cluster_selection_epsilon=cluster_selection_epsilon,
            prediction_data=True
        )

    def fit_predict(self, features: np.ndarray) -> np.ndarray:
        """
        Cluster spike features.

        Parameters
        ----------
        features : np.ndarray
            Spike features, shape (n_spikes, n_features)

        Returns
        -------
        np.ndarray
            Cluster labels (-1 for noise)
        """
        labels = self.clusterer.fit_predict(features)
        return labels

    def get_cluster_probabilities(self) -> np.ndarray:
        """Return cluster membership probabilities."""
        return self.clusterer.probabilities_

    def get_outlier_scores(self) -> np.ndarray:
        """Return outlier scores."""
        return self.clusterer.outlier_scores_
```

```python
# src/clustering/kmeans_clustering.py
"""K-means based spike clustering."""

import numpy as np
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score

class KMeansSorter:
    """Spike sorter using K-means clustering."""

    def __init__(self, n_clusters: int = None, max_clusters: int = 10):
        self.n_clusters = n_clusters
        self.max_clusters = max_clusters

    def fit_predict(self, features: np.ndarray) -> np.ndarray:
        """Cluster spike features."""
        if self.n_clusters is None:
            # Find optimal K using silhouette score
            best_k = 2
            best_score = -1

            for k in range(2, self.max_clusters + 1):
                km = KMeans(n_clusters=k, random_state=42, n_init=10)
                labels = km.fit_predict(features)
                score = silhouette_score(features, labels)
                if score > best_score:
                    best_score = score
                    best_k = k

            self.n_clusters = best_k

        km = KMeans(n_clusters=self.n_clusters, random_state=42, n_init=10)
        return km.fit_predict(features)
```

## Definition of Done

- [ ] HDBSCAN clusters without specifying K
- [ ] Cluster count reasonable (matches GT)
- [ ] >80% accuracy on tetrode data
- [ ] Outlier spikes flagged (noise cluster)
- [ ] All tests passing

## Technical Notes

- HDBSCAN advantage: no need to specify number of clusters
- Label -1 indicates noise/outlier spikes
- Consider cluster merging for over-split units

---
Part of #1 (Meta Issue)
